profiler/internal/fastdelta/fd.go

```diff
@@ -0,0 +1,817 @@
+			// location ID - repeated uint64
+			// repeated fields are packed by default
+			// see https://developers.google.com/protocol-buffers/docs/encoding#packed,
+			// but we can see both packed (bytes of concatenated PCs)
+			// or one/two single values. The Go runtime pprof encoding implementation
+			// will only pack when there are more than 2 locations for a sample
+			// (ref: https://cs.opensource.google/go/go/+/master:src/runtime/pprof/proto.go;l=527;drc=403f91c24430213b6a8efb3d143b6eae08b02ec2;bpv=1;bpt=1)
+			h.Reset()
+			switch value.WireType {
+			case codec.WireBytes:
+				err := iterPackedVarints(value.Bytes, func(id uint64) {
+					addr, ok := dc.locationIndex.Get(id)
+					if !ok {
+						return
+					}
+					binary.BigEndian.PutUint64(dc.scratch[:], addr)
+					h.Write(dc.scratch[:8])
+					dc.scratchIDs = append(dc.scratchIDs, id)
+				})
+				if err != nil {
+					return false, err
+				}
+			case codec.WireVarint:
+				addr, ok := dc.locationIndex.Get(value.Number)
+				if !ok {
+					return false, fmt.Errorf("invalid location index")
+				}
+				binary.BigEndian.PutUint64(dc.scratch[:], addr)
+				h.Write(dc.scratch[:8])
+				dc.scratchIDs = append(dc.scratchIDs, value.Number)
+			}
+			h.Sum(hash[:0])
+			dc.hashes = append(dc.hashes, *hash)
+		case recSampleValue:
+			// repeated int64
+			switch value.WireType {
+			case codec.WireBytes:
+				err := iterPackedVarints(value.Bytes, func(n uint64) {
+					values = append(values, int64(n))
+				})
+				if err != nil {
+					return false, err
+				}
+			case codec.WireVarint:
+				values = append(values, int64(value.Number))
+			}
+		case recSampleLabel:
+			var (
+				key  uint64
+				num  uint64
+				unit uint64
+			)
+			const sentinel = 0xffffffffffffffff
+			str := uint64(sentinel) // sentinel for "value is numeric, not a string"
+			err := molecule.MessageEach(codec.NewBuffer(value.Bytes), func(field int32, value molecule.Value) (bool, error) {
+				switch LabelRecordNumber(field) {
+				case recLabelKey:
+					key = value.Number
+				case recLabelStr:
+					str = value.Number
+				case recLabelNum:
+					num = value.Number
+				case recLabelNumUnit:
+					unit = value.Number
+				}
+				return true, nil
+			})
+			if err != nil {
+				return false, err
+			}
+			if !dc.strings.contains(key) {
+				return false, fmt.Errorf("invalid string index %d", key)
+			}
+			if str != sentinel && !dc.strings.contains(str) {
+				return false, fmt.Errorf("invalid string index %d", str)
+			}
+			if !dc.strings.contains(unit) {
+				return false, fmt.Errorf("invalid string index %d", unit)
+			}
+			h.Reset()
+			h.Write(dc.strings.h[uint(key)][:])
+			h.Write(dc.strings.h[uint(unit)][:])
+			binary.BigEndian.PutUint64(dc.scratch[:], num)
+			h.Write(dc.scratch[:8])
+			if str < uint64(len(dc.strings.h)) {
+				h.Write(dc.strings.h[uint(str)][:])
+			}
+			h.Sum(hash[:0])
+			dc.hashes = append(dc.hashes, *hash)
+		}
+		return true, nil
+	})
+	sort.Sort(byHash(dc.hashes))
```

Comment:
🐞 `dc.hashes` contains hashes of labels and locations. By sorting it, samples with the same set of location ids ordered in different orders will produce the same final hash. I think that's wrong.

Reply:
That makes sense to me. My intention for sorting was to consistently order the _labels_, but you're right that the call stack shouldn't be touched. I'll push a fix

Reply:
See https://github.com/DataDog/dd-trace-go/pull/1558